### PR TITLE
Fix Segment::close does not close fd issue.

### PR DESCRIPTION
**Pull request by d0ngjun**
Segment does not close the old segment file, leaving it open
permanently.

### CODE REVIEW

File: src/braft/log.cpp [hunk 1]
```diff
@@ -94,6 +94,7 @@ int Segment::create() {
         return -1;
     }
 
+    int old_fd = _fd;
```
Comment (chenzhangyi) [inline src/braft/log.cpp]: The pr is about the bug is in Segment::close, why all the modification is in create()?

Comment (d0ngjun) [reply src/braft/log.cpp]: The issue is:
Segment does not close the fd when it open a new log.
My first commit close this fd in Segment::close method, but that will break the ut.

Comment (d0ngjun) [reply src/braft/log.cpp]: Or I need close this pr and create a new one?

File: src/braft/log.cpp [hunk 2]
```diff
@@ -102,7 +103,7 @@ int Segment::create() {
     }
     LOG_IF(INFO, _fd >= 0) << "Created new segment `" << path 
                            << "' with fd=" << _fd ;
-    return _fd >= 0 ? 0 : -1;
+    return _fd >= 0 ? (old_fd > 0 ? ::close(old_fd) : 0) : -1;
```
Comment (hnwyllmm) [inline src/braft/log.cpp]: should use `old_fd >= 0` instead of `old_fd > 0` as `0` is a valid fd.

Comment (d0ngjun) [reply src/braft/log.cpp]: OK.

Another question: how to this code change?
Use forced udpate or push an individual fixup commit?

File: src/braft/log.cpp (hunk 1 above)
Comment (chenzhangyi) [reply src/braft/log.cpp]: `create()` is just another kind of `initialization`. It's not supposed that fd is valid when this function is invoked. Add a `CHECK` if you concern about this stuff.

Comment (d0ngjun) [reply src/braft/log.cpp]: This assignment does not assume fd is invalid, the check you request is under below:
old_fd >= 0 ? ::close(old_fd) : 0

Only need to close old_fd when it's valid.

**Comment by d0ngjun** [comment]
any other comments？

**Comment by chenzhangyi** [comment]
Thanks for your contribution, but this PR won't be merged.

Let me explain why:
* Segment is allowed to be accessed for read concurrently,  so that the ownership of fd has to be handled carefully:
   - Each fd is attached one and only one Segment.
   - Segment is reference counted and each `read request` acquires one ref before read the backing storage and release the ref when it finishes.
    - LogStorage owns one reference until a very segment is no longer in used(truncated or the backing node is shut).
    - fd will be closed when the reference hits 0 and Segment is destroyed. Closing the fd anywhere else might lead to potential bug.
* Segment can't be RAII since `open(2)` might fail while exception is not allowed to be used.
    - `open` means open a file which already exists on the disk.
    - `create` means create a brand new file for appending more logs.
    - Both methods should be **only invoked once** before adding the segment to LogStorage.

If you'd like to prevent all potential fd leak, make it fails when prev_fd is valid.

**Comment by chenzhangyi** [comment]
Feel free to reopen it if there's any further modification.

**Pull request closed without merging**